Add pluck/patch commands for caches and transients

petitphp opened the pull request:
Add commands for plucking and patching values from object cache and transient.

Fixes #26

petitphp commented:
I cannot reproduce the failures for the functional tests locally :
```
$ composer behat                                                                 
> run-behat-tests
...................................................................... 70
...................................................................... 140
...................................................................... 210
...................................................................... 280
...................................................................... 350
................

20 scénarios (20 succès)
366 étapes (366 succès)
4m20.27s (10.47Mb)
```

Seem's like the mu-plugins used to set up those scenarios are not loaded.

danielbachhuber commented on features/cache.feature:
```diff
@@ -210,3 +210,123 @@ Feature: Managed the WordPress object cache
     When I run `wp cache supports set_multiple`
     Then the return code should be 0
+
+  Scenario: Nested values from cache can be retrieved at any depth.
```
Can you split this to a `features/cache-pluck.feature` file?

danielbachhuber commented on features/cache.feature:
```diff
@@ -210,3 +210,123 @@ Feature: Managed the WordPress object cache
+      bar_custom
+      """
+
+  Scenario: Nested values from cache can be updated at any depth.
```
Can you split this to a `features/cache-patch.feature` file?

danielbachhuber commented on features/transient.feature:
```diff
@@ -572,3 +572,211 @@ Feature: Manage WordPress transient cache
       foo4
       """
+
+  Scenario: Nested values from transient can be retrieved at any depth.
```
Can you split this to a `features/transient-pluck.feature` file?

danielbachhuber commented on features/transient.feature:
```diff
@@ -572,3 +572,211 @@ Feature: Manage WordPress transient cache
+      baz
+      """
+
+  Scenario: Nested values from site transient can be retrieved at any depth.
```
Can you split this to a `features/transient-patch.feature` file?

danielbachhuber commented on features/cache.feature:
```diff
@@ -210,3 +210,123 @@ Feature: Managed the WordPress object cache
+      WP_CLI::add_hook( 'before_invoke:cache patch', $set_foo );
+      """
+
+    When I try `wp cache patch insert my_key fuz baz`
```
```suggestion
    When I run `wp cache patch insert my_key fuz baz`
```

danielbachhuber replied:
All of the commands that are successful should use `When I run`, not `When I try`. It looks like there are multiple occurrences.

danielbachhuber reviewed (changes requested):
> I cannot reproduce the failures for the functional tests locally :

Yeah, the tests pass locally for me too.

I don't know what the issue might be. @wp-cli/committers Any ideas?

swissspidy commented:
Unfortunately can't reproduce locally either.

Looks like `$current_value` is `null`, so the `wp_cache_get()` call in the `patch()` method is returning `null` for some reason..?

Adding some `WP_CLI::log( var_export( wp_cache_get( 'my_key' ), true ) );` calls to within `$set_foo` could help debug this.

petitphp commented:
>I manage to reproduce the failling tests locally after rebasing main in the branch, I'll look into it.

Scratch that, it was an issue with the rebase 🥲

petitphp commented:
@danielbachhuber I've split the tests into dedicated files and update them to use `When I run` if they're expected to succeed.

In https://github.com/wp-cli/cache-command/pull/89/commits/327d99a13cef3601fd8304006efaeb9396821e88 I added two debug statement to try to understand why the tests are failing in Github CI.

petitphp commented:
@danielbachhuber I was finally able to fix the failures for the functional tests. It was related to the use of STDIN to get the new value, which didn't properly handle the case where it was empty.

All tests are green on my fork.

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+	 * : Cache key.
+	 *
+	 * <key-path>...
+	 * : The name(s) of the keys within the value to locate the value to pluck.
```
```suggestion
	 * : The name(s) of the keys within the value to locate the value to patch.
```

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+	 */
+	public function pluck( $args, $assoc_args ) {
+		list($key) = $args;
+		$group     = Utils\get_flag_value( $assoc_args, 'group' );
```
```suggestion
		$group     = Utils\get_flag_value( $assoc_args, '' );
```

I think this should probably match `wp_cache_get()`'s default signature. Is there a reason you chose `'group'`?

Also, it'd be good to have functional tests covering the default group value vs. some custom group.

petitphp replied:
I don't really understand the suggestion here.

The idea was to let the user specify a custom group for the cache when plucking or patching a key. Unfortunately, I couldn't use the same pattern as the other cache commands since the number of arguments can vary, so I've used an associative argument `--group` instead.

If I understand correctly, the `get_flag_value` function take the array of associative arguments as the first parameter and the name of the argument to retrieve in this array as the second parameter. That's why I used `group` here.

You can find a functional test for the custom group in [cache-pluck.feature:29](https://github.com/wp-cli/cache-command/pull/89/files#diff-578926fa26100c7b6a371df09de2a84fa25db326468404107b313a93997fae50R29).

danielbachhuber replied:
@petitphp You are correct. I misread the code and thought `'group'` was the default value being provided for `$group`.

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+		$value = wp_cache_get( $key, $group );
+
+		if ( false === $value ) {
+			WP_CLI::halt( 1 );
```
It'd be good to have a functional test covering this scenario.

petitphp replied:
I agree this need a functional test.

This match the implementation of the option's  [`pluck` command ](https://github.com/wp-cli/entity-command/blob/main/src/Option_Command.php#L587-L589). But looking at it again here, I think it would be a better DX to use a proper error message instead of just halting the command.
```suggestion
			WP_CLI::error( "Object with key '$key' and group '$group' not found." );
```

petitphp replied:
Feature test for this case has been created.

danielbachhuber commented on src/Cache/RecursiveDataStructureTraverser.php:
```diff
@@ -0,0 +1,182 @@
+
+use UnexpectedValueException;
+
+class RecursiveDataStructureTraverser {
```
FYI - I created an issue about the duplication https://github.com/wp-cli/wp-cli/issues/5837

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+	 * ---
+	 */
+	public function pluck( $args, $assoc_args ) {
+		list($key) = $args;
```
```suggestion
		list( $key ) = $args;
```

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+	 */
+	public function patch( $args, $assoc_args ) {
+		list( $action, $key ) = $args;
+		$group                = Utils\get_flag_value( $assoc_args, 'group' );
```
Ditto comment above re: `''`.

danielbachhuber commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,168 @@ public function flush_group( $args, $assoc_args ) {
+		$traverser = new RecursiveDataStructureTraverser( $current_value );
+
+		try {
+			$traverser->$action( $key_path, $patch_value );
```
How does the new cache value get written to a persistent object cache?

petitphp replied:
As for the transient command, the cache value is updated later in the elseif condition with the `wp_cache_set` function.

danielbachhuber commented on src/Transient_Command.php:
```diff
@@ -404,6 +405,168 @@ public function list_( $args, $assoc_args ) {
+	 * : Key for the transient.
+	 *
+	 * <key-path>...
+	 * : The name(s) of the keys within the value to locate the value to pluck.
```
```suggestion
	 * : The name(s) of the keys within the value to locate the value to patch.
```

danielbachhuber commented on features/transient-patch.feature:
```diff
@@ -0,0 +1,159 @@
+        set_transient( 'my_key_2', ['foo' => ['bar' => 'baz']] );
+      };
+
+      WP_CLI::add_hook( 'before_invoke:transient patch', $set_foo );
```
Can we use `wp transient set` to set these initial transient values? The transient cache is persistent in vanilla WordPress, so we shouldn't need to hook into the command invocation in this manner.

petitphp replied:
Originally I was planing on using `wp transient set` but it is not possible to set array-like data directly like what's possible with `wp option add/update` (maybe an idea for a future PR to add this).

I've found another solution by using the `wp eval` command to execute a `set_transient`
```
Given a WP install
    And I run `wp eval "set_transient( 'my_key', ['foo' => 'bar'] );"`
    And I run `wp eval "set_transient( 'my_key_2', ['foo' => ['bar' => 'baz']] );"`
```

danielbachhuber commented on features/transient-patch.feature:
```diff
@@ -0,0 +1,159 @@
+    Then STDOUT should be:
+      """
+      Success: Value passed for transient 'my_key' is unchanged.
+      """
```
It'd be great to have a `wp transient get` to verify the value is actually saved as expected after these `wp transient patch insert` operations.

petitphp replied:
I've refactored the tests to include additional checks with `transient get`.

danielbachhuber commented on features/transient-patch.feature:
```diff
@@ -0,0 +1,159 @@
+    Then STDOUT should be:
+      """
+      Success: Value passed for transient 'my_key' is unchanged.
+      """
```
It'd be great to have a `wp transient get` to verify the value is actually saved as expected after these `wp transient patch update` operations.

petitphp replied:
I've refactored the tests to include additional checks with `transient get`.

danielbachhuber commented on features/transient-patch.feature:
```diff
@@ -0,0 +1,159 @@
+    Then STDERR should be:
+      """
+      Error: No data exists for key "foo"
+      """
```
It'd be great to have a `wp transient get` to verify the value is actually saved as expected after these `wp transient patch delete` operations.

petitphp replied:
I've refactored the tests to include additional checks with `transient get`.

danielbachhuber commented on src/Transient_Command.php:
```diff
@@ -404,6 +405,168 @@ public function list_( $args, $assoc_args ) {
+		$traverser = new RecursiveDataStructureTraverser( $current_value );
+
+		try {
+			$traverser->$action( $key_path, $patch_value );
```
How does the actual transient value get saved?

petitphp replied:
The transient value is updated later in the `elseif` condition by using the function store in the `$write_func` variable.

danielbachhuber reviewed (changes requested):
Nice work getting the tests working, @petitphp !

I left a few comments to work through.

johnbillion commented:
What's left to do on this PR?

petitphp commented:
Hi @johnbillion, sorry for the delay.

I didn't had the time to look at Daniel comments yet. I'll try to check them this week and finish up this PR for a new review.

petitphp commented:
Hi @danielbachhuber,

I've refactored the feature tests to address your feedbacks, this should be good to go.

All tests workflows using MySQL are green, the ones with SQLite are failing, but doesn't seem related to this PR specifically (eg: [Functional - WP latest on 8.2 with SQLite](https://github.com/petitphp/cache-command/actions/runs/7178714193/job/19547455581?pr=1)).

swissspidy commented:
SQLite errors are known, see #92

petitphp commented:
Hi, this PR should be ready for final review, the failling tests are related to SQLite errors and not to the changes in this PR.

swissspidy commented on composer.json:
```diff
@@ -12,7 +12,7 @@
         }
     ],
     "require": {
-        "wp-cli/wp-cli": "^2.5"
```
In this file, we can add the new commands further below in the `commands` section.

This way they will be picked up when regenerating the readme (happens automatically but you can also run `wp scaffold package-readme`)

petitphp replied:
I've added the new commands.

swissspidy commented on src/Cache_Command.php:
```diff
@@ -406,4 +409,169 @@ public function flush_group( $args, $assoc_args ) {
+		$value = wp_cache_get( $key, $group );
+
+		if ( false === $value ) {
+			WP_CLI::warning( 'No object found for the key "' . $key . '" in group "' . $group . '"' );
```
```suggestion
			WP_CLI::warning( "No object found for the key '$key' in group '$group'" );
```

Since we are using double quotes further down as well, makes it more consistent.

swissspidy commented on src/Transient_Command.php:
```diff
@@ -404,6 +405,168 @@ public function list_( $args, $assoc_args ) {
+
+		if ( $patched_value === $old_value ) {
+			WP_CLI::success( "Value passed for transient '$key' is unchanged." );
+		} elseif ( $write_func( $key, $patched_value, $expiration ) ) {
+				WP_CLI::success( "Updated transient '$key'." );
+		} else {
+			WP_CLI::error( "Could not update transient '$key'." );
+		}
```
It's easy to miss that this is the part actually saving the transient. Can we perhaps make it more verbose?

```suggestion
		} else {
			$success = $write_func( $key, $patched_value, $expiration );
			if ( $success ) {
				WP_CLI::success( "Updated transient '$key'." );
			} else {
				WP_CLI::error( "Could not update transient '$key'." );
			}
		}
```

petitphp replied:
Done in https://github.com/wp-cli/cache-command/pull/89/commits/eac9da1466c6994b6a82b78f126a4f6f488d25d0. I also updated the code for `Cache::patch` command to match.

schlessera commented:
Thanks for all the work on this, @petitphp !
